### PR TITLE
Create regex warning tests

### PR DESCRIPTION
**Pull request by JackEAllen**
## Summary of Work
Create initial unit tests for `lewis.adapters.stream.Func` in a new test file (`tests/test_stream.py`) to primarily test valid regex usage and merge ESS upstream into branch.

## Acceptance criteria
- Ensure code is readable and follows project code quality standards.
- Ensure all tests are valid and run successfully

### CODE REVIEW

File: tests/test_stream.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# *********************************************************************
+
+import unittest
+from mock import MagicMock, Mock, patch
```
Comment (JamesKingWork) [inline tests/test_stream.py]: Should we not be using unittest.mock since it moved in 3.6?

File: tests/test_stream.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    def test_func_passes_for_correct_return_argument_mappings(self):
+        self.assertTrue(Func(lambda x: 5, "([0-8]{1})", [int], int).process_request(b"7"))
+
+    def test_func_fails_for_incorrect_return_argument_mappings(self):
+        with self.assertRaises(ValueError):
+            Func(lambda x: "string", "([0-8]{1})", [int], int).process_request(b"7")
+
+    def test_process_request(self):
+        Func(lambda x: 7, "([0-8]{1})", [int], int).can_process(b"7")
```
Comment (JamesKingWork) [inline tests/test_stream.py]: Should we not be testing the return value?

File: tests/test_stream.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    def test_argument_variations_of_Func_usage(self, _, target_member, write_pattern, argument_mapping, return_mapping):
+        with self.assertRaises(RuntimeError):
+            Func(target_member, write_pattern, argument_mapping, return_mapping)
+
+    def test_argument_invalid_return_mapping_type_returns_TypeError(self):
+        with self.assertRaises(TypeError):
+            Func(lambda: 0,  "[0-9]{1}", int, 1)
+
+    def test_func_passes_for_correct_return_argument_mappings(self):
+        self.assertTrue(Func(lambda x: 5, "([0-8]{1})", [int], int).process_request(b"7"))
```
Comment (JamesKingWork) [inline tests/test_stream.py]: I may have misunderstood process_request here, but it feels like this should not return true but should return the value returned by the lambda cast to an integer?

File: tests/test_stream.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class TestFunc(unittest.TestCase):
+
+    @parameterized.expand([
+        ("valid_regex", lambda: 0, "[0-9]{1,2}", None),
+        ("valid_regex_and_argument_mapping", lambda x: 5, "([0-8]{1})", [int])
+    ])
+    def test_argument_variations_of_valid_Func_usage_without_return_mapping(self, _, target_member, write_pattern, argument_mapping):
+        self.assertTrue(Func(target_member, write_pattern, argument_mapping))
```
Comment (JamesKingWork) [inline tests/test_stream.py]: Does a constructor really return True?

File: tests/test_stream.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# *********************************************************************
+
+import unittest
+from mock import MagicMock, Mock, patch
+from lewis.adapters.stream import Func, scanf
+from parameterized import parameterized
+
+"""
```
Comment (JamesKingWork) [inline tests/test_stream.py]: Should this docstring not live in the test case?

File: tests/test_stream.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+Test Func only
+Add something to argument mappings as a dict
+may need additional mocking for Func class to operate
+test additional areas of Func after adding a few tests for invalid regex triggering an error
```
Comment (JamesKingWork) [inline tests/test_stream.py]: Are these all separate sentences? Can we add some punctuation and correct capitalisation, otherwise it's quite hard to read

File: tests/test_stream.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    def test_argument_invalid_return_mapping_type_returns_TypeError(self):
+        with self.assertRaises(TypeError):
+            Func(lambda: 0,  "[0-9]{1}", int, 1)
+
+    def test_func_passes_for_correct_return_argument_mappings(self):
+        self.assertTrue(Func(lambda x: 5, "([0-8]{1})", [int], int).process_request(b"7"))
+
+    def test_func_fails_for_incorrect_return_argument_mappings(self):
+        with self.assertRaises(ValueError):
+            Func(lambda x: "string", "([0-8]{1})", [int], int).process_request(b"7")
```
Comment (JamesKingWork) [inline tests/test_stream.py]: I find it much easier to read a construction of Func like this: `Func(target_member, write_pattern, argument_mapping, return_mapping)` than the others e.g. `Func(lambda: 0, "[0-9]{1}", int, 1)`. It lengthens the tests but could we put things like `lambda x: 5` and `"([0-8]{1})"` into explanatory variables before using them in the constructor. Especially for this line `lambda x: "string"` could explain the test a little with a name like `function_with_string_return` or `function_that_does_not_return_an_int`.

File: tests/test_stream_adapter.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+from lewis.adapters.stream import StreamHandler
+from mock import patch, MagicMock
+from parameterized import parameterized
+
+
+@patch("asynchat.async_chat")
+class TestStreamHandler(unittest.TestCase):
+    def setUp(self):
+        self.target = MagicMock()
+        self.handler = StreamHandler(MagicMock(), self.target, MagicMock)
```
Comment (JamesKingWork) [inline tests/test_stream_adapter.py]: It would be good to use explanatory fields for these MagicMocks so when someone comes to extend these tests they don't have to work out what the mocks are mocking.

File: tests/test_stream_adapter.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+    @parameterized.expand([(b"\n", "test", b"test\n"),
+                           (b"\n", b"test", b"test\n"),
+                           ("\n", "test", b"test\n"),
+                           ("\n", "test", b"test\n")])
+    @patch("asynchat.async_chat.push")
+    def test_terminator_and_replies_of_different_types_can_be_concatenated(self, terminator, message, expected, async_push, _):
+        self.target.out_terminator = terminator
+        self.handler.unsolicited_reply(message)
+        self.assertEqual(expected, async_push.call_args[0][0])
```
Comment (JamesKingWork) [inline tests/test_stream_adapter.py:24]: Is there not a nice mock assert like assert_called_with we can use here? If not can we write one as a helper method of this test?

**Review by JamesKingWork — DISMISSED** [review]
A couple of questions about why we are or are not testing certain return values and some requests to make the tests more readable so they can act as documentation

File: tests/test_stream.py (hunk 4 above)
Comment (JackEAllen) [reply tests/test_stream.py]: No, True is not returned. I used True here lazily as what is returned is just an instance of Func at a given memory address so True was simply validating that anything but an error or None was returned. I have switched this out with `assertIsInstance` of Func now.

File: tests/test_stream.py (hunk 1 above)
Comment (JackEAllen) [reply tests/test_stream.py]: Have removed entirely as not actually used. Added initially under the assumption MagicMock would be needed to test Func

File: tests/test_stream.py (hunk 6 above)
Comment (JackEAllen) [reply tests/test_stream.py]: Thanks for noticing this one. Have since removed as they were simply left over comments to help me figure out how to test initially and should not be placed into production.

File: tests/test_stream.py (hunk 3 above)
Comment (JackEAllen) [reply tests/test_stream.py]: Process request returns an integer so you understood correctly. I was being lazy again here using similar justification to use True as was used for `test_argument_variations_of_valid_Func_usage_without_return_mapping()`. I have since changed this to assert equals to compare against returned integer.

File: tests/test_stream.py (hunk 7 above)
Comment (JackEAllen) [reply tests/test_stream.py]: I have created a setup method for this and placed the occasional variable inside unitests where not re-used.
Initially chose not to use variables as I felt that this added unnecessary code which couldn't be used in parameterized. Maybe going forward, could look into swapping unnittest out for pytest to allow for using something like fixtures instead, but not really that high or a priority. 
With test naming, I'm currently of two minds about this. I agree with your point, but also want to keep test naming from being too verbose in case parameterized is used to add more edge cases later down the line to test more than just a string etc. 
I would argue that because the unit tests are so small in themselves, what is not described in the test name should easily be extracted from just reading the test itself.

File: tests/test_stream.py (hunk 2 above)
Comment (JackEAllen) [reply tests/test_stream.py]: Thanks, missed out actually asserting the bool returned here. Have added an assert now.

File: tests/test_stream_adapter.py (hunk 9 above)
Comment (JackEAllen) [reply tests/test_stream_adapter.py:24]: I believe this one may have been written by Dom, but I can't actually remember. I've modified it slightly to add an assert_called_with. I've left in the previous line in case I've misunderstood your comment and will remove it your happy with the new line.

File: tests/test_stream_adapter.py (hunk 8 above)
Comment (JackEAllen) [reply tests/test_stream_adapter.py]: I've changes this slightly to use variables if I understand your comment correctly.

**Review by JamesKingWork — APPROVED** [review]
Looks good!
